### PR TITLE
N-dimensional empty tensors: indexing, factories, reductions.

### PR DESCRIPTION
**Pull request by gchanan**
This PR implements and tests N-dimensional empty tensors for indexing, factories, and reductions if compiled with -DUSE_TH_SIZE_ZERO_DIM.

Still remaining to add:
1) TensorShape functions
2) Simple linear algebra functions (matrix multiply variants)
3) Other functions that operate over a dimension (but don't reduce).

### CODE REVIEW

File: aten/src/THC/generic/THCTensorMath.cu [hunk 1]
```diff
@@ -396,9 +396,11 @@ accreal THCTensor_(trace)(THCState *state, THCTensor *src_) {
 
 void THCTensor_(linspace)(THCState *state, THCTensor *r_, real a, real b, int64_t n) {
   THCAssertSameGPU(THCTensor_(checkGPU)(state, 1, r_));
-  THArgCheck(n > 1 || (n == 1 && (a == b)), 3, "invalid number of points");
+  // NumPy allows you to pass different points even if n <= 1 -- should we?
+  THArgCheck(n > 1 || ((n == 0 || n == 1) && (a == b)), 3, "invalid number of points");
   if (THCTensor_(nElement)(state, r_) != n) THCTensor_(resize1d)(state, r_, n);
-  if (n == 1) THCTensor_(fill)(state, r_, a);
+  if (n == 0) {
+  } else if (n == 1) THCTensor_(fill)(state, r_, a);
```
Comment (colesbury) [inline aten/src/THC/generic/THCTensorMath.cu]: I found this a little hard to read. Empty if statements are awkward so I usually put a comment to show that it's intentional. Something like:

```
if (n == 0) {
 // skip
} else if (n == 1) {
 THCTensor_(fill)(state, r_, a);
} else {
  ...
}
```

**Review by colesbury — COMMENTED** [review]
What's the `FIXME: USE_TH_SIZE_ZERO_DIM`?

File: aten/src/ATen/native/Embedding.cpp [hunk 2]
```diff
@@ -71,6 +71,7 @@ Tensor embedding_sparse_backward(
 
   // check if all our grad come from padding_idx
   if (grad.numel() == 0) {
+    // FIXME: USE_TH_SIZE_ZERO_DIM
```
[the remark this answers is not in the document]
Comment (gchanan) [reply aten/src/ATen/native/Embedding.cpp]: No, the comment is referring to sparse tensor; this PR only fixes dense tensors.

File: aten/src/THC/generic/THCTensorMath.cu (hunk 1 above)
Comment (gchanan) [reply aten/src/THC/generic/THCTensorMath.cu]: sure, good suggestion.

**Review by colesbury — APPROVED** [review]

**Pull request closed without merging**